src/MySqlConnector/ColumnReaders/BitColumnReader.cs

```diff
@@ -0,0 +1,61 @@
+namespace MySqlConnector.ColumnReaders;
+using System.Buffers.Text;
+using MySqlConnector.Protocol;
+using MySqlConnector.Protocol.Payloads;
+using MySqlConnector.Protocol.Serialization;
+using MySqlConnector.Utilities;
+
+internal sealed class BitColumnReader : IColumnReader
+{
+	internal static BitColumnReader Instance { get; } = new BitColumnReader();
+
+	public object ReadValue(ReadOnlySpan<byte> data, ColumnDefinitionPayload columnDefinition)
+	{
+		if ((columnDefinition.ColumnFlags & ColumnFlags.Binary) == 0)
+		{
+			// when the Binary flag IS NOT set, the BIT column is transmitted as MSB byte array
+			ulong bitValue = 0;
+			for (int i = 0; i < data.Length; i++)
+				bitValue = bitValue * 256 + data[i];
+			return bitValue;
+		}
+		else if (columnDefinition.ColumnLength <= 5 && data.Length == 1 && data[0] < (byte) (1 << (int) columnDefinition.ColumnLength))
+		{
+			// a server bug may return the data as binary even when we expect text: https://github.com/mysql-net/MySqlConnector/issues/713
+			// in this case, the data can't possibly be an ASCII digit, so assume it's the binary serialisation of BIT(n) where n <= 5
+			return (ulong) data[0];
+		}
+		else
+		{
+			// when the Binary flag IS set, the BIT column is transmitted as text
+			return !Utf8Parser.TryParse(data, out ulong value, out var bytesConsumed) || bytesConsumed != data.Length ? throw new FormatException() : value;
+		}
+	}
+
+	public int ReadInt32(ReadOnlySpan<byte> data, ColumnDefinitionPayload columnDefinition)
+	{
+		if ((columnDefinition.ColumnFlags & ColumnFlags.Binary) == 0)
+		{
+			// when the Binary flag IS NOT set, the BIT column is transmitted as MSB byte array
+			ulong bitValue = 0;
+			for (int i = 0; i < data.Length; i++)
+				bitValue = bitValue * 256 + data[i];
+			return checked((int) bitValue);
+		}
+		else if (columnDefinition.ColumnLength <= 5 && data.Length == 1 && data[0] < (byte) (1 << (int) columnDefinition.ColumnLength))
+		{
+			// a server bug may return the data as binary even when we expect text: https://github.com/mysql-net/MySqlConnector/issues/713
+			// in this case, the data can't possibly be an ASCII digit, so assume it's the binary serialisation of BIT(n) where n <= 5
+			return checked((int) data[0]);
+		}
+		else
+		{
+			// when the Binary flag IS set, the BIT column is transmitted as text
+			if (!Utf8Parser.TryParse(data, out long value, out var bytesConsumed) || bytesConsumed != data.Length)
+			{
+				throw new FormatException();
+			}
+			return checked((int) value);
+		}
```

Comment:
There's some significant code duplication happening here. This should be refactored into a helper method.

Will fix in https://github.com/mysql-net/MySqlConnector/pull/1332/commits/34bf3aa111a71dc57354989ec60a5f8987fd8ab8.